raw-ipa-lib/src/user.rs

```diff
@@ -64,30 +70,47 @@ impl User {
 
     fn point_from_matchkey(mk: &[u8; 32]) -> RistrettoPoint {
         // Note that ristretto wants 64 bytes of input; also we don't know if the input is uniform.
-        // TODO: Consider salting this input somehow (with the origin, perhaps).
+        // TODO: Consider salting this input somehow (with the provider, perhaps).
         //       The caveat being that anything we do needs to be standardized.
         RistrettoPoint::hash_from_bytes::<Sha512>(&mk[..])
     }
 
-    pub fn set_matchkey(&mut self, origin: &str, mk: &[u8; 32]) {
+    pub fn set_matchkey(&mut self, provider: &str, mk: &[u8; 32]) {
         let m = Self::point_from_matchkey(mk);
         let emk = self.threshold_key.encrypt(m, &mut thread_rng());
-        self.encrypted_match_keys.insert(String::from(origin), emk);
+        self.encrypted_match_keys
+            .insert(String::from(provider), emk);
     }
 
-    /// Create an encrypted matchkey for the identified origin.
+    /// Create an encrypted matchkey for the identified provider.
+    /// # Panics
+    /// If the provider name is >= 256 bytes.
     #[must_use]
-    pub fn encrypt_matchkey(&self, origin: &str) -> Ciphertext {
+    pub fn encrypt_matchkey(&self, provider: &str) -> Ciphertext {
         let mut rng = thread_rng();
         // TODO: determine if we need to hide the timing sidechannel here.
-        if let Some(emk) = self.encrypted_match_keys.get(origin) {
-            self.threshold_key.rerandomise(*emk, &mut rng)
-        } else {
-            Ciphertext::from((
-                RistrettoPoint::random(&mut rng),
-                RistrettoPoint::random(&mut rng),
-            ))
-        }
+        let emk = self
+            .encrypted_match_keys
+            .get(provider)
+            .copied()
+            .unwrap_or_else(|| {
+                let p_bytes = provider.as_bytes();
+                // This method of generating `info` doesn't need to be standardized.
+                // We're just treating HKDF as a PRG.
+                let mut info = Vec::with_capacity(256);
+                info.push(
+                    u8::try_from(p_bytes.len()).expect("provider names should be <256 bytes"),
+                );
+                info.extend_from_slice(p_bytes);
+                let mut mk = [0; 32];
+                Hkdf::<Sha512>::from_prk(&self.fallback_prk)
+                    .unwrap() // prk came from Hkdf<Sha512>
+                    .expand(&info, &mut mk)
+                    .unwrap(); // length is valid
+                let m = Self::point_from_matchkey(&mk);
+                self.threshold_key.encrypt(m, &mut thread_rng())
+            });
```

Comment:
Wow. That's a lot more complex than what I had imagined would be necessary - but I'm sure there's a good reason for all of this I don't understand.

I had imagined we would just call something like:
```
self.encrypted_match_keys
    .insert(String::from(provider), RistrettoPoint::random(&mut rng));
```